Factorization in lower precision

geoffroyleconte opened the pull request:
Choose the precision of the factorization with `Tf`.
Remove type parameterization of solve functions.

amontoison commented on src/LimitedLDLFactorizations.jl:
```diff
@@ -145,20 +145,22 @@ with the permutation vector `P`.
                    The total memory allocated is nnz(T) + n * `memory`, where
                    `T` is the strict lower triangle of A and `n` is the size of `A`.
-- `α::Tv=Tv(0)`: initial value of the shift in case the incomplete LDLᵀ
+- `α::Number=Tv(0)`: initial value of the shift in case the incomplete LDLᵀ
```
```suggestion
- `α::Number=0`: initial value of the shift in case the incomplete LDLᵀ
```

amontoison commented on src/LimitedLDLFactorizations.jl:
```diff
@@ -132,7 +132,7 @@ function LimitedLDLFactorization(
 
 """
-    LLDL = LimitedLDLFactorization(T, P; memory = 0, α = 0.0)
+    LLDL = LimitedLDLFactorization(T, P; memory = 0, α = 0.0, Tf = eltype(T))
```
```suggestion
    LLDL = LimitedLDLFactorization(T, P; memory = 0, α = 0, Tf = eltype(T))
```

amontoison commented on src/LimitedLDLFactorizations.jl:
```diff
@@ -145,20 +145,22 @@ with the permutation vector `P`.
   P::AbstractVector{<:Integer} = amd(T),
   memory::Int = 0,
-  α::Tv = Tv(0),
+  α::Number = Tv(0),
```
```suggestion
  α::Number = 0,
```

amontoison commented on src/LimitedLDLFactorizations.jl:
```diff
@@ -381,19 +383,21 @@ Compute the limited-memory LDLᵀ factorization of `A`.
   memory::Int = 0,
-  α::Tv = Tv(0),
-  droptol::Tv = Tv(0),
+  α::Number = Tv(0),
+  droptol::Number = Tv(0),
```
```suggestion
  α::Number = 0,
  droptol::Number = 0,
```

amontoison commented on src/LimitedLDLFactorizations.jl:
```diff
@@ -709,33 +713,22 @@ function lldl_solve!(n, b, Lp, Li, Lx, D, P)
-  LLDL::LimitedLDLFactorization{T, Ti},
-  b::AbstractVector{T},
-) where {T <: Real, Ti <: Integer} =
+@inline ldiv!(LLDL::LimitedLDLFactorization, b::AbstractVector) =
```
Can you remove all `@inline`? They optimize the compiler with recent versions of Julia and we don't need them now.

tmigot commented:
Ok, so if I understand the idea is that you can compute a factorization in `Tf` type while `A` is of type `T`?
Have you checked the type stability of the example you gave with @code_warntype?

dpo commented on src/LimitedLDLFactorizations.jl:
```diff
@@ -64,44 +64,44 @@ function LimitedLDLFactorization(
+  α::Number,
   n::Int,
   nnzT::Int,
+  Tf::DataType,
```
```suggestion
  Tf::DataType = Tv
```

dpo reviewed (commented):
Looks good! Here's a simple comment.

geoffroyleconte commented:
@tmigot @dpo there were indeed some type instabilities but I fixed it with my latest commit. I added a new method in case we want a lower-precision factorization.

tmigot reviewed (approved):
Looks good to me, Thanks!

dpo reviewed (approved):
Thaks @geoffroyleconte !
